crates/sui-core/src/authority_active/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,295 @@
+            self.checkpoint_store
+                .update_highest_executed_checkpoint(&last_checkpoint)
+                .unwrap();
+            self.end_of_epoch = false;
+        }
+        // Channel closed
+    }
+
+    /// Executes all checkpoints for the current epoch. At epoch boundary,
+    /// awaits the queue of scheduled checkpoints, and returns the committe
+    /// of the next epoch.
+    pub async fn execute_checkpoints_for_epoch(
+        &mut self,
+    ) -> Option<(VerifiedCheckpoint, Vec<(AuthorityPublicKeyBytes, u64)>)> {
+        let mut finished: CheckpointExecutionBuffer = FuturesOrdered::new();
+        let task_limit = TASKS_PER_CORE * num_cpus::get();
+
+        loop {
+            tokio::select! {
+                // Check for completed workers and ratchet the highest_checkpoint_executed
+                // watermark accordingly. Note that given that checkpoints are guaranteed to
+                // be processed (added to FuturesOrdered) in seq_number order, using FuturesOrdered
+                // guarantees that we will also ratchet the watermarks in order.
+                Some(Ok((seq_num, checkpoint, next_committee))) = finished.next() => {
+                    match next_committee {
+                        None => {
+                            self.metrics.last_executed_checkpoint.set(seq_num as i64);
+                            self.checkpoint_store
+                                .update_highest_executed_checkpoint(&checkpoint)
+                                .unwrap();
+
+                            let highest_synced = self
+                                .checkpoint_store
+                                .get_highest_synced_checkpoint_seq_number().unwrap()
+                                .unwrap_or(0);
+                            // If we got here, `highest_scheduled_checkpoint` must have been set,
+                            // so safe to unwrap
+                            if self.highest_scheduled_checkpoint.unwrap() < highest_synced && !self.end_of_epoch {
+                                self.schedule_checkpoint(seq_num.saturating_add(1), &mut finished).await;
+                            }
+                        }
+                        // Last checkpoint of epoch -- Note that we must not update the
+                        // highest executed watermark until after reconfig has completed!
+                        Some(committee) => return Some((checkpoint, committee)),
+                    }
+                }
+                // Read from StateSync channel if we have capacity to schedule more checkpoints.
+                // If we're at end of epoch, skip this branch so that we can effectively await
+                // execution of all remaining checkpoints of the epoch (which should be scheduled)
+                // and call reconfig.
+                received = self.mailbox.recv(), if finished.len() < task_limit && !self.end_of_epoch => match received {
```

Comment:
We stop adding new tasks to the queue when it's from a new epoch. Would this adds unnecessary delays because we have to rely on new signals to trigger the execution? It seems to me that ideally we should keep that info somewhere so that after finishing executing all checkpoints from the current epoch, we start executing ones from the new epoch

Reply:
Good catch. I think that there is a simple thing I can change that will remove some complexity and handle edge cases such as this, along with a few others, which is to remove the the reliance on new checkpoint messages from `StateSync`.